yt/frontends/gadget/io.py

```diff
@@ -383,51 +378,45 @@ def _read_particle_coords(self, chunks, ptf):
                 yield ptype, (pos[:, 0], pos[:, 1], pos[:, 2]), hsml
             f.close()
 
-    def _read_particle_fields(self, chunks, ptf, selector):
-        data_files = set()
-        for chunk in chunks:
-            for obj in chunk.objs:
-                data_files.update(obj.data_files)
-        for data_file in sorted(data_files, key=lambda x: (x.filename, x.start)):
-            poff = data_file.field_offsets
-            tp = data_file.total_particles
-            f = open(data_file.filename, "rb")
-            for ptype, field_list in sorted(ptf.items()):
-                if tp[ptype] == 0:
-                    continue
-                if getattr(selector, "is_all_data", False):
-                    mask = slice(None, None, None)
+    def _read_particle_data_file(self, data_file, ptf, selector=None):
+        return_data = {}
+        poff = data_file.field_offsets
+        tp = data_file.total_particles
+        f = open(data_file.filename, "rb")
```

Comment:
Yes, definitely safer... but I left it as is for now because we have a follow-up PR coming that adds some better file context management and takes care of this in the process.